### PR TITLE
Refactor to typescript

### PR DESCRIPTION
**Pull request by hassankhan**
Hi there,

As part of homebridge#2266, I wanted to give this project a bit of Typescript love too. 

I've converted everything except the `lib/gen/import.js`, since I'm not quite sure if that file will continue to live on. I tried running the script, and it didn't generate as many devices as the existing file had 😕 

I've also added Jest and started writing some very basic unit tests, which I'll add on later.

### CODE REVIEW

**Comment by hassankhan** [comment]
The PR is now ready for review + testing 👍

**Comment by KhaosT** [comment]
Thanks I’ll try to take a look this weekend 🙏

**Comment by hassankhan** [comment]
Really appreciate it @KhaosT 👍.

Once you check out the branch, it'd be good to run `yarn` and `yarn build` to update the dependencies and compile the Typescript respectively. Then run `node dist/BridgedCore.js` (for example) and you should see some output in the CLI.

**Comment by KhaosT** [comment]
@hassankhan I tired to run the built version and when trying to pair with the accessory, I run into

```
/dist/lib/util/tlv.js:22
    if (data.length <= 255) {
             ^

TypeError: Cannot read property 'length' of undefined
```

File: src/lib/util/tlv.ts [hunk 1]
```diff
@@ -47,8 +40,8 @@ function encode(type, data /*, type, data, type, data... */) {
     if (arguments.length > 2) {
```
Comment (KhaosT) [inline src/lib/util/tlv.ts]: here we probably should check `args` instead of `arguments`?

Comment (KhaosT) [reply src/lib/util/tlv.ts]: Ah okay this part we also need to fix nextArgs.

```
  if (args.length >= 2) {

      // chop off the first two arguments which we already processed, and process the rest recursively
      const [ nextType, nextData, ...nextArgs ] = args;
      const remainingTLVBuffer = encode(nextType, nextData, ...nextArgs);

      // append the remaining encoded arguments directly to the buffer
      encodedTLVBuffer = Buffer.concat([encodedTLVBuffer, remainingTLVBuffer]);
    }
```

Comment (hassankhan) [reply src/lib/util/tlv.ts]: HI @KhaosT, good catch, must've missed them in all the refactoring excitement 😄! I'll address both the above issues in a commit shortly.

**Comment by hassankhan** [comment]
Thanks for taking a look at the branch, @KhaosT 👍 I was wondering if you had a "test plan" that I could use myself?

**Comment by KhaosT** [comment]
@hassankhan There isn't a formalized test plan 😅 In general for changes like this I'll probably test the following:

1. Start a fresh accessory (not paired with HomeKit before), try pair with it from iOS device, control it and make sure sample accessories work as expected.
2. Shutdown the instance, and restart it, verify iOS can re-establish secure session.

Then from homebridge side, test with one or two simple plugins, make sure API matches what original implementation is.

**Comment by hassankhan** [comment]
Hi @KhaosT, no worries, it'd be really nice to one day have some sort of integration tests, but I'm not sure how (without messing around with the iOS simulator).

So far, I've tried the following and things _seem_ to be working fine:

- `yarn && yarn build`
- `node dist/BridgedCore.js`
- Open Home app on iOS and add new Accessory
- Use pincode from `src/BridgedCore.ts`
- Once accessory is paired, try changing values on all the bridged sample accessories
- Killed the running session with Ctrl-C and then restarted it, once again tried changing values on all bridged sample accessories

So far, all of the above worked perfectly. Should I be testing `Core.ts` and `CameraCore.ts` in the same way?

I haven't yet tested Homebridge using this branch, that will be next.

**Comment by hobbyquaker** [comment]
> it'd be really nice to one day have some sort of integration tests, but I'm not sure how (without messing around with the iOS simulator).

@hassankhan I'm doing automated testing for my HAP-Nodejs based Project [homekit2mqtt](https://github.com/hobbyquaker/homekit2mqtt) without the simulator, I'm utilizing the hap-client from @forty2 https://github.com/forty2/hap-client-tool
See test output here to get an idea what can be done: https://travis-ci.org/hobbyquaker/homekit2mqtt/jobs/420789242#L3348
hth, regards, Sebastian

**Comment by hassankhan** [comment]
OMG that's awesome @hobbyquaker 🤯

**Comment by hassankhan** [comment]
Hi @KhaosT, reckon you could give the branch another go this weekend? Would massively appreciate your input 😄

**Comment by hassankhan** [comment]
Hi @KhaosT, I've updated the branch quite a bit, it seems to be working for me 👍

**Comment by KhaosT** [comment]
Thanks ^^ I can take a look on Saturday 😊

**Comment by KhaosT** [comment]
Thanks for doing this, this looks awesome!

I'm about to merge this PR. TypeScript is relatively new to me, is there anything special I need to do for npm publish? And for end package consumer, are there any additional setup they need to do to use the library?

**Comment by hassankhan** [comment]
No worries at all, @KhaosT, thanks for the support! 😄 

I've added a new package script which should automatically build it when an NPM release happens. 

I'm also looking at creating another PR for code-quality tooling as well as some contributer documentation 👍

In terms of the end-users of this project, I've only had the chance to try it with Homebridge, and with non-camera accessories only - so maybe not as thorough as I'd like. Perhaps we could reach out to some other projects and ask them to try this branch too?

In terms of breakages, the only thing I can think of is things like the `once()` helper function that were directly imported before now no longer exist at those paths, so it may also be an idea to bump the minor version.

**Comment by KhaosT** [comment]
Thanks, I'll merge it now.

I think currently `ts-node src/BridgedCore.ts` will complain about `Could not find a declaration file for module 'buffer-shims'.`; we can fix it separately.

**Comment by KhaosT** [comment]
v0.5.0 is released for this change 🎉

**Comment by hassankhan** [comment]
Thank you so much @KhaosT! Do let me know if anything comes up, more than happy to fix any issues I may have introduced 👍

**Comment by KhaosT** [comment]
@hassankhan thank you for converting the project to typescript 🙏 I think the only thing need to fix is that `buffer-shims` type definition lookup with ts-node. And maybe add instruction for people to who want to run `Core.js` directly (it's rare but I think some people use it in that way)? Otherwise I tried a few projects linking directly to `hap-nodejs` and they all function fine.

**Comment by aboulfad** [comment]
>... And maybe add instruction for people to who want to run `Core.js` directly (it's rare but I think some people use it in that way)?

Am one of those ‘rare’ people that still use Core.js but will continue using the older HAP-Node.js version until I truly understand the benefit that typescript brings to HAP-node specifically. Regardless, thanks and respect to all your effort for maintaining this project.

**Comment by aboulfad** [comment]
> v0.5.0 is released for this change 🎉

V.0.5.0 is the latest release before these changes https://github.com/KhaosT/HAP-NodeJS/tree/v0.4.50
Maybe you meant v0.5.1 which doesn’t show under releases.

**Comment by aboulfad** [comment]
> v0.5.0 is released for this change 🎉
Can’t find v0.5.0 under releases ? https://github.com/KhaosT/HAP-NodeJS/releases

**Comment by KhaosT** [comment]
@aboulfad For some reasons, `npm version` stopped tagging the release. I'll look into fixing that when I have time ^^

**Comment by aboulfad** [comment]
> @aboulfad For some reasons, `npm version` stopped tagging the release. I'll look into fixing that when I have time ^^

Thanks, no rush :-) in the meantime one can alway checkout older commits that are labelled using release tags (e.g `git checkout tags/<tag_name> -b <branch_name> `

**Comment by hassankhan** [comment]
> until I truly understand the benefit that typescript brings to HAP-node specifically.

Hi @aboulfad, I just thought I'd address your (valid) concern about TypeScript. Ultimately, to an end-user of this project, it wouldn't bring any direct benefits, at least in the short-term.

The groups of people who would see an immediate benefit are (a) HAP-NodeJS contributors, and (b) projects that use HAP-NodeJS as a core component. Thanks to most of the types now being declared (and TypeScript's code-completion functionality) you can now at a glance understand what data types are being passed around. This is _extremely_ useful for new HAP-NodeJS contributors who may never have interacted with the codebase before, and would not otherwise (easily) be able to understand those same data types. It is also very useful for any downstream projects, since any changes/additions/removals are easily communicated through the types file (although obviously they should also be communicated via documentation etc.)

> Regardless, thanks and respect to all your combined efforts for maintaining this awesome repo.

Thank you for your kind words 😃, big thanks to @KhaosT for being awesome and merging this in so quickly :+1:

**Comment by aboulfad** [comment]
Thank you @hassankhan for a nice explanation. Now you peeked my curiously, I read quickly about typescript so maybe I’ll give the master branch a chance and see how easy is to migrate to using ts. From the look of it, it should be in theory transparent. .

**Comment by n0rt0nthec4t** [comment]
I'm also one of those users who run Core.js.  Had a look this morning and now unsure what is required to migrate previously developed accessories to this new codebase... Guess now need to understand what is typescript and all that it brings

**Comment by aboulfad** [comment]
@n0rt0nthec4t , actually the impact is independent of  `Core` or `BridgedCore`. It affects any accessories developed using HAP-Node. We have the choice to stay on pre ts, or learn and embark on this journey. Check my above post with the diff between pre and post ts.

**Comment by n0rt0nthec4t** [comment]
@aboulfad Thanks. I think for the moment will stay on the pre-ts and see how things shake out. Too may accessories to update and down quite have the time to learn typescript

**Comment by OliverS79** [comment]
@KhaosT @hassankhan - Dear both,  Thanks for the TypeScript version. As I am still new to TS, I have the following challenge. I follow the readme instruction and try to run ts-node Core.ts in the src folder. However, I get the following error: "Could not find a declaration file for module 'buffer-shims'" 

`npm install @types/buffer-shims` does not work either.

Can you give me some hint on how to get this one work? Have I missed something in the readme file?

Best,
Oliver

**Comment by aboulfad** [comment]
@OliverS79 FYI, your issue is referenced twice in this thread but no mention of a fix has been delivered or a workaround.

**Comment by hassankhan** [comment]
Hi @aboulfad, @OliverS79,

I've been extremely busy with work during the week, I will try and get a PR to hopefully resolve some of these issues for you.

@aboulfad The diff link you posted above does not work for me. If I get an example of what broke, I can try and fix it 😄

**Comment by aboulfad** [comment]
@hassankhan , I fixed the differences [link](https://github.com/aboulfad/HAP-NodeJS/compare/aboulfad:v0.4.53...master#diff-f01fb3685ba13a9e762d06e8faa59228) in my above post. 
Though I never said the ts changes broke something, rather that they change the syntax somewhat for the common HAP-Node methods/functions and as a consequence existing JS accessory files will have to change. I only quickly glanced at this, but it is evident that basic JS users will have to learn a bit about TS to modify their scripts IF they want to user HAP-NodeTS.

Example: 
Using JS:
```
lightAccessory
  .addService(Service.Lightbulb, LightController.name) 
  .getCharacteristic(Characteristic.On)	
  .on('set', function(value, callback)
 ```
becomes using TS:
```
lightAccessory
  .addService(Service.Lightbulb, LightController.name) 
  .getCharacteristic(Characteristic.On)
  .on(CharacteristicEventTypes.SET, (value: CharacteristicValue, callback: CharacteristicSetCallback) => {
    LightController.setPower(value);
```

**Comment by hassankhan** [comment]
@aboulfad thanks for updating the link. 

I think I should probably change the documentation somewhat... no-one should _need_ to convert their projects to start using Typescript, you should be able to continue using existing Javascript syntax with no change in behaviour. The only time you _should_ start seeing any differences is when you (re)write your code in Typescript.

**Comment by aboulfad** [comment]
@hassankhan , my bad ... looking at the accessory examples led me to believe that one would have to make those changes. So you changed all accessory files as an example how to write an accessory using TS ?

**Comment by hassankhan** [comment]
@aboulfad Yes, unfortunately I did. Again, totally my fault for not updating the documentation - sorry for the trouble it caused you 😞

**Comment by aboulfad** [comment]
^^^ No trouble at all, it was just a simple quick diff that showed whats going on :-) And now we have examples for those that want to migrate to TS :-)

**Comment by n0rt0nthec4t** [comment]
> @KhaosT @hassankhan - I get the following error: "Could not find a declaration file for module 'buffer-shims'"
> 
> `npm install @types/buffer-shims` does not work either.

Been trying to install this new typescript beast with teh latest code base and still get the error as above.

I'm very much in two minds about this change to the code base, from an end-user perspective it seems very much retrograde step. No fast startup for Core.js(ts) amore, as ts-node takes forever to verify everything *sigh*

**Comment by hassankhan** [comment]
Hi all, sorry for the delay in responding - had some IRL issues to iron out.

@n0rt0nthec4t I'm terribly sorry for the issues you're experiencing ... I've got a branch with some documentation updates, hopefully I can help clarify things there. Do you also run your own fork of HAP-NodeJS or are you using it as a dependency?

**Comment by Supereg** [comment]
@hassankhan 
I'm using HAP-NodeJS as a dependency in my project. Get the following two error messages when I try to compile my typescript project:

```
node_modules/hap-nodejs/dist/lib/Advertiser.d.ts:1:55 - error TS7016: Could not find a declaration file for module 'bonjour-hap'. '/project-path/node_modules/bonjour-hap/index.js' implicitly has an 'any' type.
  Try `npm install @types/bonjour-hap` if it exists or add a new declaration (.d.ts) file containing `declare module 'bonjour-hap';`

1 import { BonjourHap, MulticastOptions, Service } from 'bonjour-hap';
                                                        ~~~~~~~~~~~~~

node_modules/hap-nodejs/dist/lib/HAPServer.d.ts:2:17 - error TS7016: Could not find a declaration file for module 'fast-srp-hap'. '/project-path/node_modules/fast-srp-hap/index.js' implicitly has an 'any' type.
  Try `npm install @types/fast-srp-hap` if it exists or add a new declaration (.d.ts) file containing `declare module 'fast-srp-hap';`

2 import srp from 'fast-srp-hap';
                  ~~~~~~~~~~~~~~


Found 2 errors.
```

**Comment by n0rt0nthec4t** [comment]
> Hi all, sorry for the delay in responding - had some IRL issues to iron out.
> 
> @n0rt0nthec4t I'm terribly sorry for the issues you're experiencing ... I've got a branch with some documentation updates, hopefully I can help clarify things there. Do you also run your own fork of HAP-NodeJS or are you using it as a dependency?

Just run as a dependency for my accessories, previously node Core.js would start up instantly but not with the typescript thing, takes around 1+ min on a rpi zero. So for me, this is very much backwards code base.  Will be staying on 0.4.x atm

**Comment by Supereg** [comment]
@n0rt0nthec4t this is the thing with typescript. Before every run it transpiles to javascript and then gets executed (when running with ts-node). You could potentially compile it once and then only run the compiled java script code. Although on every update you need to compile again.

**Comment by n0rt0nthec4t** [comment]
> @n0rt0nthec4t this is the thing with typescript. Before every run it transpiles to javascript and then gets executed (when running with ts-node). You could potentially compile it once and then only run the compiled java script code. Although on every update you need to compile again.

Guessing then the compile to JS would only be needed if you update the HAP-NodeJS code, and if your accessories are still in JS, then no issue?

Perhaps some more documentation/guidance is needed for this?

**Pull request merged**